src/mcp/server.ts

```diff
@@ -46,17 +46,80 @@ export type ToolSchema<Input extends z.Schema> = {
 
 export type ToolHandler = (toolName: string, params: any) => Promise<ToolResponse>;
 
+export interface InitializeInfo {
+  clientVersion: Implementation;
+  roots?: ListRootsResult;
+}
+
 export interface ServerBackend {
   name: string;
   version: string;
-  initialize?(server: Server): Promise<void>;
+  initialize?(info: InitializeInfo): Promise<void>;
   tools(): ToolSchema<any>[];
   callTool(schema: ToolSchema<any>, parsedArguments: any): Promise<ToolResponse>;
   serverClosed?(): void;
+
+  onChangeProxyTarget?: (target: string, options: any) => void;
 }
 
 export type ServerBackendFactory = () => ServerBackend;
 
+export class ServerBackendSwitcher implements ServerBackend {
+  private _target: ServerBackend;
+  private _initializeInfo?: InitializeInfo;
+
+  constructor(private readonly _targetFactories: Record<string, (options: any) => ServerBackend>) {
+    const defaultTargetFactory = this._targetFactories[''];
+    this._target = defaultTargetFactory({});
+    this._target.onChangeProxyTarget = this._handleChangeProxyTarget;
+  }
+
+  _handleChangeProxyTarget = (name: string, options: any) => {
+    const factory = this._targetFactories[name];
+    if (!factory)
+      throw new Error(`Unknown target: ${name}`);
+    const target = factory(options);
+    this.switch(target).catch(logUnhandledError);
+  };
+
+  async switch(target: ServerBackend) {
+    const old = this._target;
+    old.onChangeProxyTarget = undefined;
+    this._target = target;
+    this._target.onChangeProxyTarget = this._handleChangeProxyTarget;
```

Comment:
Why do we have to assign `onChangeProxyTarget` to the target while we could just handle that particular tool call in this class ant spare all other backends from the knowledge about `onChangeProxyTarget`? Basically similar to how we switch between context factories but just intercept the backend switch call here.